File: clients/android/sample/src/main/kotlin/dev/spectre/sample/SampleHostDelegate.kt
package dev.spectre.sample

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.util.Log
import android.widget.Toast
import dev.spectre.core.SpValue
import dev.spectre.core.SpectreActionResponse
import dev.spectre.core.SpectreDestination
import dev.spectre.core.SpectreHostDelegate
import dev.spectre.core.SpectreRequest
import dev.spectre.core.stringify
import kotlinx.coroutines.delay

/**
 * サンプルアプリのホスト実装。
 *
 * ここが「SDK がホストアプリに要求するもの」のすべて (docs/architecture.md §6)。
 * 実アプリではこの中で既存のネットワーク層・ルーティング・計測基盤に繋ぐ。
 *
 * サンプルではサーバを立てずに応答を模擬している。重要なのは
 * **論理エンドポイント名 -> 実リクエスト の解決がホスト側にある**という構造で、
 * これによりドキュメントに内部 URL や認証情報が載ることがない。
 */
class SampleHostDelegate(
    private val context: Context,
    private val onEvent: (String) -> Unit = {},
) : SpectreHostDelegate {

    /** ホストが受け付ける論理エンドポイント。ここにない名前は実行されない。 */
    private val registeredEndpoints = setOf("cart.add", "settings.save")

    /** ホストが受け付ける URL のドメイン。ここにないホストは開かない。 */
    private val allowedUrlHosts = setOf("example.com", "www.example.com")

    /** ホストが提供する機能。ドキュメントからは `host` アクションで呼ばれる。 */
    private val registeredHostActions = setOf("share")

    override suspend fun performRequest(request: SpectreRequest): SpectreActionResponse {
        if (request.endpoint !in registeredEndpoints) {
            Log.w(TAG, "未登録のエンドポイント: ${request.endpoint}")
            return SpectreActionResponse.failure(
                "ENDPOINT_NOT_REGISTERED",
                "エンドポイント '${request.endpoint}' は登録されていません",
            )
        }

        onEvent("→ ${request.method} ${request.endpoint} ${request.body.stringify()}")
        delay(600) // ネットワーク往復の体感を出すための待ち

        return when (request.endpoint) {
            "cart.add" -> {
                // サーバが返した値を data にマージする形の応答
                val qty = (request.body as? SpValue.Obj)?.entries?.get("qty")?.stringify() ?: "1"
                SpectreActionResponse(
                    ok = true,
                    state = SpValue.Obj(mapOf("cartCount" to SpValue.Str(qty))),
                )
            }

            "settings.save" -> {
                // 3回に1回失敗させて onError 側の経路も確認できるようにする
                failureCounter++
                if (failureCounter % 3 == 0) {
                    SpectreActionResponse.failure("SERVER_BUSY", "サーバが混み合っています")
                } else {
                    SpectreActionResponse(ok = true)
                }
            }

            else -> SpectreActionResponse(ok = true)
        }
    }

    override fun navigate(destination: SpectreDestination): Boolean {
        // サンプルには遷移先の画面がないので、受け取ったことだけを見せる
        onEvent("navigate ${destination.mode} ${destination.screen ?: destination.route ?: ""}")
        return false
    }

    override suspend fun performHostAction(name: String, params: SpValue): SpValue? {
        if (name !in registeredHostActions) {
            Log.w(TAG, "未登録の host アクション: $name")
            throw IllegalArgumentException("host アクション '$name' は登録されていません")
        }
        onEvent("host:$name")
        if (name == "share") {
            val obj = params as? SpValue.Obj
            val text = obj?.entries?.get("text")?.stringify().orEmpty()
            val url = obj?.entries?.get("url")?.stringify().orEmpty()
            val intent = Intent(Intent.ACTION_SEND).apply {
                type = "text/plain"
                putExtra(Intent.EXTRA_TEXT, listOf(text, url).filter { it.isNotEmpty() }.joinToString("\n"))
            }
            context.startActivity(Intent.createChooser(intent, null).addFlags(Intent.FLAG_ACTIVITY_NEW_TASK))
        }
        return null
    }

    override fun track(event: String, properties: SpValue) {
        onEvent("track $event ${properties.stringify()}")
        Log.d(TAG, "track $event ${properties.stringify()}")
    }

    override fun openUrl(url: String, mode: String): Boolean {
        val host = runCatching { Uri.parse(url).host }.getOrNull()
        if (host == null || host !in allowedUrlHosts) {
            // アロウリスト外は開かない。ドキュメントは公開物なので、
            // 任意の URL を開けるとフィッシングの経路になる。
            Log.w(TAG, "アロウリスト外の URL: $url")
            Toast.makeText(context, "許可されていないURLです: $url", Toast.LENGTH_SHORT).show()
            return false
        }
        onEvent("openUrl $url")
        runCatching {
            context.startActivity(
                Intent(Intent.ACTION_VIEW, Uri.parse(url)).addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            )
        }
        return true
    }

    private var failureCounter = 0

    private companion object {
        const val TAG = "SpectreSample"
    }
}

File: clients/ios/SampleApp/Sources/SampleHostDelegate.swift
import Foundation
import UIKit
import SpectreCore

/// サンプルアプリのホスト実装。
///
/// ここが「SDK がホストアプリに要求するもの」のすべて (docs/architecture.md §6)。
/// 実アプリではこの中で既存のネットワーク層・ルーティング・計測基盤に繋ぐ。
///
/// サンプルではサーバを立てずに応答を模擬している。重要なのは
/// **論理エンドポイント名 -> 実リクエスト の解決がホスト側にある**という構造で、
/// これによりドキュメントに内部 URL や認証情報が載ることがない。
/// Android の `SampleHostDelegate.kt` と対になっている。
final class SampleHostDelegate: SpectreHostDelegate {

    /// ホストが受け付ける論理エンドポイント。ここにない名前は実行されない。
    private let registeredEndpoints: Set<String> = ["cart.add", "settings.save"]

    /// ホストが受け付ける URL のドメイン。ここにないホストは開かない。
    private let allowedURLHosts: Set<String> = ["example.com", "www.example.com"]

    /// ホストが提供する機能。ドキュメントからは `host` アクションで呼ばれる。
    private let registeredHostActions: Set<String> = ["share"]

    private let onEvent: (String) -> Void
    private var failureCounter = 0

    init(onEvent: @escaping (String) -> Void = { _ in }) {
        self.onEvent = onEvent
    }

    func performRequest(_ request: SpectreRequest) async throws -> SpectreActionResponse {
        guard registeredEndpoints.contains(request.endpoint) else {
            return .failure(
                code: "ENDPOINT_NOT_REGISTERED",
                message: "エンドポイント '\(request.endpoint)' は登録されていません"
            )
        }

        onEvent("→ \(request.method) \(request.endpoint) \(request.body.stringify())")
        // ネットワーク往復の体感を出すための待ち
        try? await Task.sleep(nanoseconds: 600_000_000)

        switch request.endpoint {
        case "cart.add":
            let qty = request.body["qty"]?.stringify() ?? "1"
            return SpectreActionResponse(ok: true, state: .object(["cartCount": .string(qty)]))

        case "settings.save":
            // 3回に1回失敗させて onError 側の経路も確認できるようにする
            failureCounter += 1
            if failureCounter % 3 == 0 {
                return .failure(code: "SERVER_BUSY", message: "サーバが混み合っています")
            }
            return SpectreActionResponse(ok: true)

        default:
            return SpectreActionResponse(ok: true)
        }
    }

    func navigate(to destination: SpectreDestination) -> Bool {
        // サンプルには遷移先の画面がないので、受け取ったことだけを見せる
        onEvent("navigate \(destination.mode) \(destination.screen ?? destination.route ?? "")")
        return false
    }

    func performHostAction(name: String, params: SpValue) async throws -> SpValue? {
        guard registeredHostActions.contains(name) else {
            throw SampleHostError.unregisteredAction(name)
        }
        onEvent("host:\(name)")
        if name == "share" {
            let text = params["text"]?.stringify() ?? ""
            let url = params["url"]?.stringify() ?? ""
            await MainActor.run {
                let items: [Any] = [text, url].filter { !($0 as String).isEmpty }
                let activity = UIActivityViewController(activityItems: items, applicationActivities: nil)
                Self.topViewController()?.present(activity, animated: true)
            }
        }
        return nil
    }

    func track(event: String, properties: SpValue) {
        onEvent("track \(event) \(properties.stringify())")
    }

    func openURL(_ url: String, mode: String) -> Bool {
        guard let parsed = URL(string: url), let host = parsed.host, allowedURLHosts.contains(host) else {
            // アロウリスト外は開かない。ドキュメントは公開物なので、
            // 任意の URL を開けるとフィッシングの経路になる。
            onEvent("openUrl 拒否 (アロウリスト外) \(url)")
            return false
        }
        onEvent("openUrl \(url)")
        Task { @MainActor in UIApplication.shared.open(parsed) }
        return true
    }

    @MainActor
    private static func topViewController() -> UIViewController? {
        let scene = UIApplication.shared.connectedScenes.first as? UIWindowScene
        var top = scene?.windows.first(where: \.isKeyWindow)?.rootViewController
        while let presented = top?.presentedViewController { top = presented }
        return top
    }
}

enum SampleHostError: LocalizedError {
    case unregisteredAction(String)

    var errorDescription: String? {
        switch self {
        case .unregisteredAction(let name):
            return "host アクション '\(name)' は登録されていません"
        }
    }
}
